In [67]:
import numpy as np
import pandas as pd
from siuba import *
from calitp import *
import intake
import data_prep
import fuzzymatcher
pd.options.display.max_rows = 250

In [68]:
df = data_prep.load_grantprojects()
vehicles = data_prep.load_vehiclesdata()

/home/jovyan/data-analyses/5311/data_prep.py:51: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [69]:
# Grouping vehicle types together: not working in function
Automobiles = ['Automobile','Automobiles (Service)','Sports Utility Vehicle']
Bus = ['Bus','Over-the-road Bus']
Vans = ['Van','Trucks and other Rubber Tire Vehicles (Service)','Minivan','Cutaway',]

In [70]:
    def replace_modes(row):
        if row.vehicle_type in Automobiles:
            return "Rail"
        elif row.vehicle_type in Bus:
            return "Bus"
        else:
            return "Vans"
vehicles["vehicle_groups"] = vehicles.apply(lambda x: replace_modes(x), axis=1)

In [71]:
vehicles.head(3)

agency    city state legacy_ntd_id  \
1362  Fresno County Rural Transit Agency  Fresno    CA      9R02-025   
1363  Fresno County Rural Transit Agency  Fresno    CA      9R02-025   
1364  Fresno County Rural Transit Agency  Fresno    CA      9R02-025   

          ntd_id                                  organization_type  \
1362  9R02-91007  Independent Public Agency or Authority of Tran...   
1363  9R02-91007  Independent Public Agency or Authority of Tran...   
1364  9R02-91007  Independent Public Agency or Authority of Tran...   

       reporter_type  primary_uza_population  agency_voms  \
1362  Rural Reporter                       0           77   
1363  Rural Reporter                       0           77   
1364  Rural Reporter                       0           77   

                                         vehicle_type  years_old:  _13_15  \
1362                                              Van         NaN       0   
1363  Trucks and other Rubber Tire Vehicles (Service)         NaN       0   
1364                                          Minivan         NaN       0   

      _16_20  _21_25  _26_30  _31_60  _60+  total_vehicles  \
1362       0       0       0       0     0              18   
1363       0       0       0       0     0               3   
1364       0       0       0       0     0               4   

      average_age_of_fleet__in_years_  average_lifetime_miles_per_vehicle  \
1362                         2.555556                                 0.0   
1363                              NaN                                 NaN   
1364                        11.000000                                 0.0   

      _0_12 vehicle_groups  
1362     18           Vans  
1363      3           Vans  
1364      4           Vans

# Look at unique rural agencies in vehicles

In [72]:
#subset for crosswalk
vehicles_crosswalk = vehicles[['agency','ntd_id']]

In [73]:
#drop duplicates for crosswalk
vehicles_crosswalk = vehicles_crosswalk.drop_duplicates()

In [100]:
f'{len(vehicles_crosswalk)} total agencies.'

'52 total agencies.'

In [74]:
vehicles_crosswalk.sort_values('agency')

agency      ntd_id
6668      Alpine County Local Transportation Commission  9R02-91116
4637                     Amador Regional Transit System  9R02-91000
5797                           Calaveras Transit Agency  9R02-99442
6451                                     City of Arcata  9R02-91018
5788                                      City of Arvin  9R02-91027
6001                                     City of Auburn  9R02-91032
6460                            City of California City  9R02-91111
6665  City of Chowchilla, dba: Chowchilla Area Transit   9R02-91071
5786       City of Corcoran, dba: Corcoran Area Transit  9R02-91002
6004                                     City of Dinuba  9R02-91040
5149                     City of Dixon, dba: Readi-Ride  9R02-91041
6666                       City of Escalon, dba: eTrans  9R02-91078
5155        City of Eureka, dba: Eureka Transit Service  9R02-91093
6455                                  City of Guadalupe  9R02-91043
6760                                  City of McFarland  9R02-91110
6759         City of Needles, dba: Needles Area Transit  9R02-91020
6663                                       City of Ojai  9R02-91058
5997               City of Ridgecrest, dba: Ridgerunner  9R02-91006
6449     City of Rio Vista, dba: Rio Vista Delta Breeze  9R02-91014
5156          City of Shafter, dba: Shafter Dial-A-Ride  9R02-91120
6000                                    City of Solvang  9R02-91028
6457               City of Taft, dba: Taft Area Transit  9R02-91066
6459                                   City of Woodlake  9R02-91079
5792                       Colusa County Transit Agency  9R02-91112
4452  County of Sacramento Municipal Services Agency...  9R02-90216
5151   County of Siskiyou, dba: Siskiyou County Transit  9R02-91048
2045                   Eastern Sierra Transit Authority  9R02-91062
1362                 Fresno County Rural Transit Agency  9R02-91007
6243                              Glenn Transit Service  9R02-91088
4296                         Humboldt Transit Authority  9R02-91036
2197                              Kern Regional Transit  9R02-91059
2727                             Lake Transit Authority  9R02-91053
6006                      Lassen Transit Service Agency  9R02-91098
5399  Madera County, dba: Madera County Connection T...  9R02-91005
5401              Mariposa County Transit, dba: Mari-Go  9R02-91082
3199                        Mendocino Transit Authority  9R02-91047
6447                        Modoc Transportation Agency  9R02-91008
4454                    Morongo Basin Transit Authority  9R02-91090
3907  Mountain Area Regional Transit Authority, dba:...  9R02-91012
4158                     Nevada County Transit Services  9R02-91095
5800                   Palo Verde Valley Transit Agency  9R02-99454
6461            Plumas County Transportation Commission  9R02-91119
5405                    Redwood Coast Transit Authority  9R02-91097
3791                              San Benito County LTA  9R02-91009
6241            Sierra County Transportation Commission  9R02-91038
4300                                      Tehama County  9R02-91089
5791                                    Town of Truckee  9R02-91101
6238                                     Trinity County  9R02-91035
3202                         Tulare County Area Transit  9R02-91055
3553                            Tuolumne County Transit  9R02-91057
6670                                     Wasco, City of  9R02-99426
4994       Yosemite Area Regional Transportation System  9R02-91070

# Using Organizations CSV (Put in Script Later)
* CSV with Cal ITP id and gtfs stuff
* Throw all the organizations without NTP ID
* Use vehicle dataset to see which one is missing

In [75]:
GCS_FILE_PATH = "gs://calitp-analytics-data/data-analyses/5311 /"
FILE_NAME = "organizations-all_organizations.csv"
organizations = pd.read_csv(f"{GCS_FILE_PATH}{FILE_NAME}")

In [76]:
#put in script
organizations = to_snakecase(organizations)

In [77]:
#put in script, keep ony certain cols
organizations = organizations[['name','ntp_id','itp_id','gtfs_schedule_status','#_services_w__complete_rt_status',
       '#_fixed_route_services_w__static_gtfs',
       'complete_static_gtfs_coverage__1=yes_', 'complete_rt_coverage',
       '>=1_gtfs_feed_for_any_service__1=yes_',
       '>=_1_complete_rt_set__1=yes_']]

In [78]:
#put in script, renaming columns
organizations = organizations.rename(columns = {'name':'agency','ntp_id':'ntd_id'})

In [79]:
organizations.shape

(781, 10)

### Delete all records without any NTD ID 
* Went from 781 rows to 199

In [80]:
organizations2 = organizations.dropna(subset=['ntd_id'])

In [81]:
organizations2.shape

(199, 10)

In [82]:
organizations_crosswalk = organizations2[['agency','ntd_id', 'itp_id']]

#  Only 26 records match between vehicles & GTFS status on agency & NTD ID.
* Using left join, organization crosswalk on the L. Vehicles on the R. 

In [108]:
df_joined1 = organizations_crosswalk[organizations_crosswalk.ntd_id.notna()].merge(vehicles_crosswalk.drop_duplicates(), on=['agency', 'ntd_id'],  how='left', 
                                           validate = "1:1", indicator=True)

In [109]:
df_joined1._merge.value_counts()

left_only     173
both           26
right_only      0
Name: _merge, dtype: int64

### Tiltering out for only agencies with both...

In [110]:
df_joined_both = df_joined1.loc[df_joined1['_merge'] == 'both']

In [86]:
df_joined_both=df_joined_both.drop_duplicates()

In [87]:
df_joined_both

agency      ntd_id  itp_id _merge
3                  Amador Regional Transit System  9R02-91000    11.0   both
17                                 City of Arcata  9R02-91018    18.0   both
19                                  City of Arvin  9R02-91027    21.0   both
21                                 City of Auburn  9R02-91032    23.0   both
31                        City of California City  9R02-91111    51.0   both
45                                 City of Dinuba  9R02-91040    93.0   both
57                              City of Guadalupe  9R02-91043   129.0   both
73                              City of McFarland  9R02-91110   197.0   both
81                                   City of Ojai  9R02-91058   231.0   both
99                                City of Solvang  9R02-91028   312.0   both
115                              City of Woodlake  9R02-91079   371.0   both
116                  Colusa County Transit Agency  9R02-91112    74.0   both
119              Eastern Sierra Transit Authority  9R02-91062    99.0   both
122            Fresno County Rural Transit Agency  9R02-91007   117.0   both
127                    Humboldt Transit Authority  9R02-91036   135.0   both
131                        Lake Transit Authority  9R02-91053   159.0   both
132                 Lassen Transit Service Agency  9R02-91098   162.0   both
138                   Mendocino Transit Authority  9R02-91047   198.0   both
141               Morongo Basin Transit Authority  9R02-91090   212.0   both
156               Redwood Coast Transit Authority  9R02-91097   261.0   both
173       Sierra County Transportation Commission  9R02-91038   304.0   both
182                                 Tehama County  9R02-91089   334.0   both
183                               Town of Truckee  9R02-91101   340.0   both
185                                Trinity County  9R02-91035   344.0   both
186                    Tulare County Area Transit  9R02-91055   346.0   both
195  Yosemite Area Regional Transportation System  9R02-91070   374.0   both

#### Filtering out for agencies that appeared in vehicles but not in GTFS data set

In [104]:
df_vehicles_only = df_joined1.loc[df_joined1['_merge'] == 'left_only']

In [105]:
df_vehicles_only

agency      ntd_id  itp_id  \
0                                      Access Services       90157     1.0   
1                Alameda-Contra Costa Transit District       90014     4.0   
2                                        Alpine County  9R02-91116     9.0   
4                       Anaheim Transportation Network       90211    14.0   
5                    Antelope Valley Transit Authority       90121    16.0   
6                                  Bishop Paiute Tribe       99268    40.0   
7                                  Blue Lake Rancheria       99292    42.0   
8              Butte County Association of Governments       90208    47.0   
9                                     Calaveras County  9R02-91063    50.0   
10             California Department of Transportation        9R02     NaN   
11                        California Vanpool Authority       90230    53.0   
12                             Chemehuevi Indian Tribe       99316    64.0   
13                    City and County of San Francisco       90015   282.0   
14                                City of Agoura Hills       90246     2.0   
15                                    City of Alhambra       90247     6.0   
16                                     City of Arcadia       90044    17.0   
18                                     City of Artesia       90300    20.0   
20                                  City of Atascadero       90194    22.0   
22                                      City of Avalon       90249    24.0   
23                                       City of Azusa       90250    26.0   
24                                City of Baldwin Park       90251    29.0   
25                                        City of Bell       90252    35.0   
26                                City of Bell Gardens       90253    36.0   
27                                  City of Bellflower       90254    37.0   
28                               City of Beverly Hills       90255    39.0   
29                                     City of Burbank       90257    45.0   
30                                   City of Calabasas       90256    49.0   
32                                   City of Camarillo       90163    54.0   
33                                      City of Carson       90258    57.0   
34                                    City of Cerritos       90259    63.0   
35                                  City of Chowchilla  9R02-91071    65.0   
36                                   City of Claremont       90296     NaN   
37                                    City of Commerce       90043    75.0   
38                                     City of Compton       90260    77.0   
39                                    City of Corcoran  9R02-91002    78.0   
40                                      City of Corona       90052    79.0   
41                                      City of Covina       90261    85.0   
42                                      City of Cudahy       90262    86.0   
43                                       City of Davis       90167     NaN   
44                                      City of Delano       90238    91.0   
46                                       City of Dixon  9R02-91040    94.0   
47                                      City of Downey       90263    95.0   
48                                      City of Duarte       90264    97.0   
49                                    City of El Monte       90265   102.0   
50                                   City of Elk Grove       90205     NaN   
51                                      City of Eureka  9R02-91093   108.0   
52                                   City of Fairfield       90092   110.0   
53                                      City of Folsom       90220     NaN   
54                                      City of Fresno       90027   116.0   
55                                    City of Glendale       99423   120.0   
56                                    City of Glendora       90266   1

# How many agencies do not have ITP ID..
* 25  don't have IDS?

In [112]:
df_joined2 = vehicles_crosswalk.merge(organizations_crosswalk.drop_duplicates(), on=['agency','ntd_id'], validate = "1:1", how='left', indicator=True)
df_joined2._merge.value_counts()

left_only     26
both          26
right_only     0
Name: _merge, dtype: int64

In [114]:
df_joined2.sort_values('_merge')

agency      ntd_id  itp_id  \
16                     City of Dixon, dba: Readi-Ride  9R02-91041     NaN   
23       City of Corcoran, dba: Corcoran Area Transit  9R02-91002     NaN   
27                           Calaveras Transit Agency  9R02-99442     NaN   
21              Mariposa County Transit, dba: Mari-Go  9R02-91082     NaN   
20  Madera County, dba: Madera County Connection T...  9R02-91005     NaN   
19          City of Shafter, dba: Shafter Dial-A-Ride  9R02-91120     NaN   
18        City of Eureka, dba: Eureka Transit Service  9R02-91093     NaN   
17   County of Siskiyou, dba: Siskiyou County Transit  9R02-91048     NaN   
28                   Palo Verde Valley Transit Agency  9R02-99454     NaN   
41               City of Taft, dba: Taft Area Transit  9R02-91066     NaN   
29               City of Ridgecrest, dba: Ridgerunner  9R02-91006     NaN   
37                        Modoc Transportation Agency  9R02-91008     NaN   
12  County of Sacramento Municipal Services Agency...  9R02-90216     NaN   
44            Plumas County Transportation Commission  9R02-91119     NaN   
36                              Glenn Transit Service  9R02-91088     NaN   
9                      Nevada County Transit Services  9R02-91095     NaN   
8   Mountain Area Regional Transit Authority, dba:...  9R02-91012     NaN   
7                               San Benito County LTA  9R02-91009     NaN   
6                             Tuolumne County Transit  9R02-91057     NaN   
46  City of Chowchilla, dba: Chowchilla Area Transit   9R02-91071     NaN   
47                       City of Escalon, dba: eTrans  9R02-91078     NaN   
48      Alpine County Local Transportation Commission  9R02-91116     NaN   
2                               Kern Regional Transit  9R02-91059     NaN   
49                                     Wasco, City of  9R02-99426     NaN   
38     City of Rio Vista, dba: Rio Vista Delta Breeze  9R02-91014     NaN   
50         City of Needles, dba: Needles Area Transit  9R02-91020     NaN   
0                  Fresno County Rural Transit Agency  9R02-91007   117.0   
40                                  City of Guadalupe  9R02-91043   129.0   
35            Sierra County Transportation Commission  9R02-91038   304.0   
42                                   City of Woodlake  9R02-91079   371.0   
43                            City of California City  9R02-91111    51.0   
45                                       City of Ojai  9R02-91058   231.0   
39                                     City of Arcata  9R02-91018    18.0   
34                                     Trinity County  9R02-91035   344.0   
25                                    Town of Truckee  9R02-91101   340.0   
32                                     City of Dinuba  9R02-91040    93.0   
1                    Eastern Sierra Transit Authority  9R02-91062    99.0   
3                              Lake Transit Authority  9R02-91053   159.0   
4                         Mendocino Transit Authority  9R02-91047   198.0   
5                          Tulare County Area Transit  9R02-91055   346.0   
10                         Humboldt Transit Authority  9R02-91036   135.0   
11                                      Tehama County  9R02-91089   334.0   
33                      Lassen Transit Service Agency  9R02-91098   162.0   
13                    Morongo Basin Transit Authority  9R02-91090   212.0   
15       Yosemite Area Regional Transportation System  9R02-91070   374.0   
22                    Redwood Coast Transit Authority  9R02-91097   261.0   
24                                      City of Arvin  9R02-91027    21.0   
26                       Colusa County Transit Agency  9R02-91112    74.0   
30                                    City of Solvang  9R02-91028   312.0   
31                                     City of Auburn  9R02-91032    23.0   
14                     Amador Regional Transit System  9R02-91000    11.0   
51                                  City of

In [90]:
df_joined2.isna().sum()

agency     0
ntd_id     0
itp_id    26
_merge     0
dtype: int64

# 5311 Data Crosswalk Test
* Comparing the agency names in 5311 against Vehicle Data to ensure everything is the same using fuzzy matcher https://pbpython.com/record-linking.html

In [91]:
df.head(2)

grant_fiscal_year funding_program             grant_number  project_year  \
0               2011    Section 5311  CA-18-X047 | 0012000083          2016   
1               2011    Section 5311  CA-18-X047 | 0012000083          2016   

     organization_name        upin                   description     ali  \
0  City of Chowchilla   BCG0000228          Operating Assistance  300902   
1        Madera County  BCG0000283  Buy <30-Ft Bus For Expansion  111304   

  contract_number  allocationamount  encumbered_amount  expendedamount  \
0    64BO17-00368           53221.0           114511.0        53221.00   
1    64BC17-00408          110663.0           110663.0       101352.02   

   activebalance  closedoutbalance project_status  
0           0.00                 0           Open  
1        9310.98                 0           Open

In [92]:
#put in script, filter funding program
df = to_snakecase(df)
subset = ['Section 5311', '5310 Exp', '5310 Trad', '5311(f) Cont', 'CMAQ (FTA 5311)',
       'Section 5311(f)','5311(f) Round 2']
df = df[df.funding_program.isin(subset)]

In [93]:
#subset for crosswalk
crosswalk_5311 = df[['organization_name']]
crosswalk_5311 = crosswalk_5311.drop_duplicates()

### Testing join on our crosswalk subsets
* Left join on GTFS & Vehicle crosswalk
* All pd.merge should have `validate="1:1"` or `validate="m:1"`


In [94]:
left_on = ["organization_name"]
right_on = ["agency"]

In [95]:
matched_results = fuzzymatcher.fuzzy_left_join(df_joined_both,crosswalk_5311, right_on, left_on)

### Looking at best match score...
* Tulare, Amador, and Trinity need to be filtered out.
* Either use dictionary or merge black cat back in
* Filter out for left only in organization crosswalk

In [96]:
matched_results.sort_values('best_match_score')

best_match_score __id_left __id_right  \
28          0.011467   24_left   44_right   
0           0.042136    0_left   43_right   
27          0.107495   23_left   72_right   
25          0.187038   21_left   69_right   
8           0.339274    4_left   14_right   
5           0.371082    1_left    4_right   
7           0.371082    3_left   13_right   
13          0.371082    9_left   25_right   
6           0.371082    2_left   12_right   
11          0.371082    7_left   19_right   
10          0.371082    6_left   18_right   
9           0.371082    5_left    5_right   
18          0.423514   14_left    9_right   
19          0.429384   15_left   45_right   
21          0.455322   17_left   49_right   
12          0.470893    8_left    6_right   
14          0.470893   10_left   29_right   
26          0.470966   22_left   70_right   
24          0.521156   20_left   65_right   
15          0.563976   11_left   30_right   
23          0.599631   19_left   57_right   
22          0.612896   18_left   51_right   
16          0.621225   12_left   38_right   
20          0.636715   16_left   46_right   
17          0.741619   13_left    2_right   
29          0.847407   25_left  177_right   

                                          agency      ntd_id  itp_id _merge  \
28                    Tulare County Area Transit  9R02-91055   346.0   both   
0                 Amador Regional Transit System  9R02-91000    11.0   both   
27                                Trinity County  9R02-91035   344.0   both   
25                                 Tehama County  9R02-91089   334.0   both   
8                        City of California City  9R02-91111    51.0   both   
5                                 City of Arcata  9R02-91018    18.0   both   
7                                 City of Auburn  9R02-91032    23.0   both   
13                               City of Solvang  9R02-91028   312.0   both   
6                                  City of Arvin  9R02-91027    21.0   both   
11                             City of McFarland  9R02-91110   197.0   both   
10                             City of Guadalupe  9R02-91043   129.0   both   
9                                 City of Dinuba  9R02-91040    93.0   both   
18                    Humboldt Transit Authority  9R02-91036   135.0   both   
19                        Lake Transit Authority  9R02-91053   159.0   both   
21                   Mendocino Transit Authority  9R02-91047   198.0   both   
12                                  City of Ojai  9R02-91058   231.0   both   
14                              City of Woodlake  9R02-91079   371.0   both   
26                               Town of Truckee  9R02-91101   340.0   both   
24       Sierra County Transportation Commission  9R02-91038   304.0   both   
15                  Colusa County Transit Agency  9R02-91112    74.0   both   
23               Redwood Coast Transit Authority  9R02-91097   261.0   both   
22               Morongo Basin Transit Authority  9R02-91090   212.0   both   
16              Eastern Sierra Transit Authority  9R02-91062    99.0   both   
20                 Lassen Transit Service Agency  9R02-91098   162.0   both   
17            Fresno County Rural Transit Agency  9R02-91007   117.0   both   
29  Yosemite Area Regional Transportation System  9R02-91070   374.0   both   

                                organization_name  
28        Kings County Area Public Transit Agency  
0                           Kern Regional Transit  
27   Trinity County Department of Transportation   
25                   Tehama County Transit Agency  
8                         City of California City  
5                                  City of Arcata  
7                                  City of Auburn  
13                                City of Solvang  
6                                   City of Arvin  
11                              City of McFarland  
10                              City of Guadalupe  
9                             

In [97]:
f'we have a total of {matched_results.agency.nunique()} unique organization names...'

'we have a total of 26 unique organization names...'